### PR TITLE
fix(security): report repository-integrity findings on fork PRs

### PR DESCRIPTION
**Pull request by TheZupZup**
## Summary

The repository-integrity scanner correctly **blocks** external PRs #513, #514 and #515 and attributes every finding to `bd48e57df55ed0cde25170e4baa53314bd2187e8` — including on #513, whose final diff against `main` is empty. But no sticky comment was ever posted, so contributors saw a red check with no explanation.

This PR fixes the trusted reporter only. Detection semantics, the history scanner, and the scanner workflow's read-only trust model are unchanged.

**Do not merge #513/#514/#515.** They are incident reproductions and must stay blocked.

## Root cause

The reporter's job-level condition required `github.event.workflow_run.pull_requests[0] != null`. GitHub delivers that array **empty for every fork-originated run**, so the job was skipped before the artifact could be inspected — all three reporter runs for #513/#514/#515 show `conclusion: skipped`. Two downstream reads shared the assumption: the comment step interpolated `pull_requests[0].number`, and the renderer required `len(pull_requests) == 1`.

## Two trust problems, not one

Fixing the skip exposed a second, deeper issue. They are separate and both are addressed.

**1. Which pull request does this report describe?** — identity binding.
**2. Is this report honest?** — verdict provenance.

### 1. Identity binding

The scanner workflow file is contributor-controlled at the merge ref, so the artifact's `pr_number` is a **claim**, never authority.

| Field | Trust |
|---|---|
| `workflow_run.event` / `.name` / `.path` / `.repository.full_name` | GitHub-set — pins the producer to the scanner's immutable file path |
| `workflow_run.id` | GitHub-set — the artifact is downloaded with `run-id` pinned to it |
| `workflow_run.head_sha`, `.head_repository.full_name`, `.head_branch` | GitHub-set from the PR that started the run — **cannot be forged from inside the run** |
| artifact `pr_number` / `head_sha` / `head_repository` | Untrusted claims — reconciled against the above |
| API `commits/{sha}/pulls` on the head repo, `pulls.get(N)` on this repo | Derived, authoritative |

`scripts/resolve_repository_integrity_pr.py` accepts the claim only when the run is the scanner's own; the artifact's head SHA and head repository equal the run's; exactly one PR carries that head commit *and* head repository; and the live PR — fetched by number **from this repository** — agrees on number, base repository, head repository, head branch and head SHA.

**The association query targets the head repository.** A fork's head commit is not in the base repository's commit list. Confirmed against the live API with #513's head:

```
GET /repos/TheZupZup/Linthra/commits/a9ecac5b…/pulls   -> []
GET /repos/Borhan2004/Linthra/commits/a9ecac5b…/pulls  -> [#513]
```

Asking the base repository sees zero candidates and fails closed, so fork recovery would resolve nothing. Asking a contributor-owned repository grants nothing: the name is GitHub-set, the PR itself is still read from **this** repository, and every candidate stays constrained by base repository, head repository, head SHA and claimed number. The response is read whole — the endpoint pages at 30 and advertises the rest via `Link: rel="next"`, so an unpaginated read could both miss the PR and hide a second match from the ambiguity check. Accumulation is bounded, a non-terminating chain fails closed, and the next URL must be on the API host.

### 2. Verdict provenance — the artifact's findings are never published

Identity binding proves *which* PR a report describes. It cannot prove the report is honest. A fork contributor can keep the scanner's name and path, drop the step that loads the trusted checker, and upload an artifact whose identity fields are genuine and whose `findings` list is empty. Every field identity binding checks would be honest; only the verdict is forged — and publishing it would put a **CLEAN comment under `github-actions[bot]`**, lending contributor-authored content the authority of the maintainer-controlled reporter. Extending the reporter to forks is what exposed this: before this branch, fork runs were skipped entirely.

So the verdict is re-derived. `scripts/recompute_repository_integrity_report.py` runs the checker from the trusted default-branch checkout over the same commit range, with base SHA, head SHA and PR author taken from the **live pull request** rather than the artifact. The uploaded report is still downloaded and identity-bound, and is compared against the re-derivation; a divergence is reported and **the re-derived result wins**, because refusing to comment on a mismatch would let a forged artifact suppress the finding it was forged to hide.

**No PR code is executed and no PR tree is checked out.** The checker is a pure git-object analyser — `show`, `ls-tree`, `rev-list`, `merge-base`, `diff`, and it reads `.gitignore` as a *blob* rather than shelling to `git check-ignore` — so objects are fetched into the trusted checkout through this repository's own `refs/pull/N/head`, never from the contributor's remote, and analysed as data. The working tree stays on the default branch; the workflow contains no `checkout`, `switch` or `worktree` of PR content, asserted by test. The checker is invoked as an argument vector, never a shell string.

## Fail-closed behaviour

Malformed artifact, missing binding, mismatch, more than one plausible PR, zero plausible PRs, unexpected workflow path, unexpected repository, or a checker that cannot evaluate the range → job fails, no comment. A head that moved on since the scan is superseded, so an out-of-order run leaves the newer verdict standing.

One case is deliberately not a failure: the guard job does not run on draft PRs, so those runs upload no artifact. That is told apart from every other reason a report can be missing — transport error, expired artifact, permissions failure, a scan that died before uploading — by reading the run's artifact list and the guard job's conclusion from the Actions API **before** downloading, **and** confirming the PR's own `draft` flag. A skipped guard job is not proof of a draft: the condition that skips it lives in the contributor-controlled workflow file. Nothing in the reporter tolerates failure — there is no `continue-on-error` anywhere in it.

## Preserved protections

Trusted scanner path binding · schema validation · head SHA binding · head repository binding · head branch binding (new) · PR number binding · stale/out-of-order protection · field-length limits · finding-count limits · total rendered comment budget · Markdown/HTML/mention escaping · one sticky bot comment · BLOCKED / REVIEW REQUIRED / CLEAN.

Unchanged: scanner is `contents: read` with no secrets and no write token; reporter is a separate trusted `workflow_run` workflow running from the default branch; no `pull_request_target` anywhere.

## Review findings addressed

Seven automated findings; six genuine and fixed, one refuted. Each verified against a primary source before acting.

| # | Finding | Disposition |
|---|---|---|
| 1 | Artifact suppression too broad (`continue-on-error` swallowed every download failure) | Fixed — Actions-API skip/failure distinction |
| 2 | Branch allowlist rejected valid git branch names | Fixed — confirmed with `git check-ref-format` that `feature+test`, `feature@2`, `feature/ümlaut` are valid and were rejected; now bounded opaque data |
| 3 | Association response not paginated | Fixed — confirmed the live endpoint returns `Link: rel="next"` |
| 4 | "`head_sha` is the synthetic merge revision" | **Refuted** — runs [32902865997](https://github.com/TheZupZup/Linthra/actions/runs/32902865997) and [32887057547](https://github.com/TheZupZup/Linthra/actions/runs/32887057547) both report the raw PR head; the merge revision is `github.sha`, a different field |
| 5 | A skipped guard job is not proof of a draft PR | Fixed — the PR's own `draft` flag is now required |
| 6 | Artifact contents unauthenticated → forged CLEAN | Fixed — verdict re-derived in trusted code (above) |
| 7 | Author allowlist rejected `dependabot[bot]` | Fixed — same root cause as #2; now bounded opaque data |

Findings 2 and 7 shared a root cause worth noting: over-constraining a GitHub-provided value that is only ever used as data, turning cosmetic hardening into a functional outage for legitimate inputs. Both are now bounded rather than shaped.

## Test plan

- [x] `check_repository_integrity_test.py` — **73 passed** (was 72; +1 for the #513 reproduction)
- [x] `render_repository_integrity_comment_test.py` — **17 passed** (was 15)
- [x] `resolve_repository_integrity_pr_test.py` — **34 passed** (new file)
- [x] `recompute_repository_integrity_report_test.py` — **10 passed** (new file)
- [x] `python3 -m py_compile` on all four scripts
- [x] All `.github/workflows/*.yml` parse; both `github-script` bodies pass `node --check`
- [x] `git diff --check` clean
- [x] `check_pr_security_surface.py` on this diff: exit 0, sensitive-not-blocked
- [x] `check_repository_integrity.py` on this diff: passes as owner, blocks as external
- [x] End-to-end: contaminated commit, cleanup, Update-branch merge → **zero changed files**, scanner still exits 1, renderer produces `BLOCKED` naming the historical commit
- [x] Old vs new against a fixture modelling the real #513 API behaviour — old fails closed with `0 pull requests match this run's head`; new resolves #513 via the fork endpoint

**Adversarial test for finding 6:** a fork PR introduces `.vscode/tasks.json`, drops `.vscode/` from `.gitignore`, cleans up, and uploads a forged CLEAN report with correct identity. The test asserts the forged artifact *does* pass identity validation and *would* render CLEAN, then that re-derivation finds 3 findings and renders **BLOCKED** naming the offending commit.

The identity test mock previously returned the fork PR from the *base* repository path, which does not happen in production and is what let the association bug through. It now routes by repository the way GitHub actually behaves.

Further regression coverage: forged `pr_number`; head SHA / head repository / head branch mismatch; artifact from the wrong run; wrong scanner workflow path; stale head after synchronize; fork BLOCKED → CLEAN in one sticky comment; hostile strings cannot form an API path; bot-form logins accepted; multi-page association responses concatenate; oversized and non-terminating pagination fail closed; pagination cannot be walked off the API host; no `pull_request_target` in any workflow; scanner has no write token; reporter never checks out PR head; only a skipped-and-draft PR counts as nothing to report; an honest clean PR still renders CLEAN; an unusable range errors rather than reading as clean.

## #513 preservation

Intact. The end-to-end test reproduces the incident's exact finding shapes — `.gitignore` dropping the required `.vscode/` entry, all five `.vscode/` files, `.vscode/tasks.json` invoking an asset as executable, and a text-only `public/fonts/fa-solid-400.woff2` — introduced in one commit, cleaned up, then merged with trusted `main` so the final diff is empty. Every path is still attributed to that single contaminated commit, every finding is `blocked`, and the Update-branch merge is not blamed. (The fixture SHA differs from `bd48e57…` because the history is built fresh; the property under test is identical.)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01FZ4Ks1czg9foX6bEehxyj3

### CODE REVIEW

**Comment by TheZupZup** [comment]
CI is green on `d17e42b` — all 14 PR checks succeeded or skipped, including `Repository integrity guard tests`, which now runs the new `resolve_repository_integrity_pr_test.py` suite.

One observation worth recording, because it's live evidence for the `continue-on-error` change flagged in the description: this PR opened as a draft, so `Repository integrity guard` was skipped by its `if: !github.event.pull_request.draft` condition and uploaded no findings artifact. The reporter on `main` then fired and **failed**: [run 32889355326](https://github.com/TheZupZup/Linthra/actions/runs/32889355326). That is the pre-existing draft-PR failure described in the last section of the description, reproducing on this very PR.

It does not block anything here — `workflow_run` runs are not PR checks, which is why the PR is green — but it is the case the new "Require a findings artifact" gate exists to handle. It can only take effect once merged, since `workflow_run` workflows always execute from the default branch.

To be explicit about what that gate does and does not change: no comment is written in this case either before or after, and enforcement is entirely in the scanner's own check, which is unaffected. The gate only stops the reporter job going red for a run that has nothing to report. Every other artifact condition — malformed, unbindable, mismatched, ambiguous — still fails the job loudly.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: scripts/resolve_repository_integrity_pr.py [hunk 1]
```diff
@@ -0,0 +1,457 @@
...
+
+def resolve(event: object, payload: object, repository: str) -> Binding:
+    """Bind the artifact to the triggering run, or refuse."""
+    run = trusted_run(event, repository)
+    pr_number, head_sha, head_repository = artifact_identity(payload)
+
+    # The artifact must describe the exact head this run scanned. An artifact
+    # produced by any other run cannot satisfy both of these.
+    if head_sha != run.head_sha:
+        raise BindingError("report head SHA does not match the triggering run")
```
[the remark this answers is not in the document]
Comment (TheZupZup) [reply scripts/resolve_repository_integrity_pr.py:266]: Not applying this one — the premise is refuted by this repository's own runs.

`workflow_run.head_sha` on a `pull_request` run is the **raw PR head commit**, not a synthetic merge revision. Two independent checks against the live API:

**This PR (same-repo), scanner run [32902865997](https://github.com/TheZupZup/Linthra/actions/runs/32902865997):**
```
event:    pull_request
path:     .github/workflows/repository-integrity.yml
head_sha: 68c58bc2f6c396858923356e144acf0c6d648cfe
```
That is exactly the commit pushed to this branch, and the run's `head_commit.message` is that commit's own subject ("fix: read the commit-association response whole…"). A synthetic merge revision would carry neither.

**PR #513 (fork), scanner run [32887057547](https://github.com/TheZupZup/Linthra/actions/runs/32887057547):**
```
run head_sha:      a9ecac5b6a94159c590ccb51e09f5a00ac14f796
PR #513 head.sha:  a9ecac5b6a94159c590ccb51e09f5a00ac14f796
```
Identical. So on the exact fork PR this change targets, the artifact's `github.event.pull_request.head.sha` and the run's `head_sha` are the same value and the binding matches rather than "always hits this mismatch".

The underlying distinction is real but concerns a different field: `github.sha` / `GITHUB_SHA` — what `actions/checkout` resolves by default — *is* the merge revision for `pull_request` events. `workflow_run.head_sha` in the `workflow_run` event payload is not. The `flutter-ci-fixer.yml` comment cited is about the checked-out revision; note that this scanner does not rely on the default checkout at all, since it pins `ref: ${{ github.event.pull_request.head.sha }}` explicitly.

Were the premise true, the same mismatch would already break the pre-existing same-repo path merged in #522, which has been posting sticky comments successfully — including the CLEAN comment on this PR.

Marking resolved on the evidence above rather than changing the binding, since binding to a merge commit's parent would be strictly weaker than binding to a SHA GitHub sets directly. Happy to reopen if there is a counter-example run where the two values differ.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by TheZupZup** [comment]
**`Build debug APK` is failing on `fb1617a`, and it is not this PR's.** [Job log](https://github.com/TheZupZup/Linthra/actions/runs/32903390436/job/97981996739):

```
Download action repository 'actions/setup-java@v5' …
##[warning]Failed to download action … Error: Name or service not known
                                        (internal-api.service.iad.github.net:443)
##[error]Failed to download archive … after 3 attempts.
```

The job died in **Prepare all required actions** — a DNS failure fetching the `actions/setup-java` tarball from GitHub's own internal host. It never reached checkout, Gradle, or any build step, so nothing from this branch executed.

Three things place it outside this PR:

- `fb1617a` differs from `68c58bc` by **a 12-line Python comment block and nothing else**.
- The PR touches eight files — two workflows, two scripts, three test files, one doc. No Android, Gradle, Java, Kotlin, Dart, or `pubspec` file is in the diff.
- The same job passed on this PR's own earlier head `d17e42b`.

I don't have permission to re-run it (`rerun-failed-jobs` returns 403 for this integration), and the failure isn't something that can be made robust from inside this PR's scope — the fix would be in `android-debug-apk.yml`, which this change deliberately doesn't touch. So I'm not pushing anything for it.

Noting it once here rather than going quiet. The PR stays watched, and I'll re-check until it's green, merged, or closed. A maintainer re-run of that one job should clear it; if it reproduces identically on a re-run, that would make it a real signal worth looking at separately from this PR.

All other checks on `fb1617a` are green, including `Repository integrity guard`, `Repository integrity guard tests`, `Security surface guard`, and `Secret & privacy scan`.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: .github/workflows/repository-integrity-reporter.yml [hunk 2]
```diff
@@ -32,32 +42,92 @@ jobs:
...
+            const artifacts = await github.paginate(
+              github.rest.actions.listWorkflowRunArtifacts, { ...common, per_page: 100 });
+            if (artifacts.some(a => a.name === 'repository-integrity-findings')) {
+              core.setOutput('present', 'true');
+              return;
+            }
+            const jobs = await github.paginate(
+              github.rest.actions.listJobsForWorkflowRun, { ...common, per_page: 100 });
+            const guard = jobs.find(job => job.name === 'Repository integrity guard');
+            if (guard?.conclusion !== 'skipped') {
```
[the remark this answers is not in the document]
Comment (TheZupZup) [reply .github/workflows/repository-integrity-reporter.yml:79]: Valid, and the most consequential finding on this PR. Fixed in `a5efc88`.

You're right that identity binding proves *which* PR a report describes but not that the report is honest, and that every field the binder checks would be genuine in the forged case. Worth stating plainly: extending the reporter to forks is what exposed this — before this branch, fork runs were skipped, so a forged verdict could never reach an untrusted PR.

**The artifact's `findings` are no longer the source of any published verdict.** The reporter re-derives them itself in `scripts/recompute_repository_integrity_report.py`, running the checker from the trusted default-branch checkout over the same commit range. Its inputs — base SHA, head SHA, PR author — are read from the live pull request via the API and carried on the binding, never taken from the artifact.

On your "never execute PR-controlled code" constraint: nothing is executed and no PR tree is checked out. The checker is a pure git-object analyser — `show`, `ls-tree`, `rev-list`, `merge-base`, `diff`, and it reads `.gitignore` as a *blob* rather than shelling to `git check-ignore` — so the objects are fetched into the trusted checkout via this repository's own `refs/pull/N/head` (never from the contributor's remote) and analysed as data. The working tree stays on the default branch; the workflow contains no `checkout`, `switch` or `worktree` of PR content, asserted by test. The checker is invoked as an argument vector rather than a shell string, and every binding field it receives is re-validated on read.

The uploaded report is still downloaded and identity-bound, and is now compared against the re-derivation. A divergence is reported and **the re-derived result wins** — refusing to comment on a mismatch would let a forged artifact suppress the very finding it was forged to hide.

The adversarial regression test reproduces the attack end to end: a fork PR that introduces `.vscode/tasks.json`, drops `.vscode/` from `.gitignore`, cleans up, and uploads a forged CLEAN report with correct identity. It asserts the forged artifact *does* pass identity validation and *would* render CLEAN, then that re-derivation finds 3 findings and renders **BLOCKED** naming the offending commit:

```
::warning::the uploaded artifact reported 0 finding(s) but the trusted checker
re-derived 3; the re-derived result is authoritative
```

Also covered: an honest clean PR still renders CLEAN (no false positives from re-derivation), an unusable range errors rather than reading as an empty finding list, and the uploaded artifact is never the file passed to the renderer.

Fork-PR binding, stale-head protection and fail-closed behaviour are unchanged. Suites: 73 / 17 / 32 / 9 new.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: .github/workflows/repository-integrity-reporter.yml [hunk 3]
```diff
@@ -32,32 +42,131 @@ jobs:
...
+            const run_id = Number.parseInt(process.env.RUN_ID, 10);
+            if (!Number.isSafeInteger(run_id) || run_id < 1) {
+              throw new Error('invalid triggering run id');
+            }
+            const common = { owner: context.repo.owner, repo: context.repo.repo, run_id };
+            const artifacts = await github.paginate(
+              github.rest.actions.listWorkflowRunArtifacts, { ...common, per_page: 100 });
+            if (artifacts.some(a => a.name === 'repository-integrity-findings')) {
+              core.setOutput('present', 'true');
+              return;
```
[the remark this answers is not in the document]
Comment (TheZupZup) [reply .github/workflows/repository-integrity-reporter.yml:74]: Valid, and the most consequential finding on this PR. Fixed in `a5efc88`. (I first replied to this on the neighbouring draft-verification thread by mistake — this is the thread it belongs on.)

You're right that identity binding proves *which* PR a report describes but not that the report is honest: in the forged case every field the binder checks is genuine, and only the verdict is fabricated. Worth stating plainly — extending the reporter to forks is what exposed this. Before this branch, fork runs were skipped, so a forged verdict could never reach an untrusted PR.

**The artifact's `findings` are no longer the source of any published verdict.** The reporter re-derives them itself in `scripts/recompute_repository_integrity_report.py`, running the checker from the trusted default-branch checkout over the same commit range. Its inputs — base SHA, head SHA, PR author — come from the live pull request via the API and are carried on the binding, never taken from the artifact.

No PR code is executed and no PR tree is checked out. The checker is a pure git-object analyser (`show`, `ls-tree`, `rev-list`, `merge-base`, `diff`, and it reads `.gitignore` as a *blob* rather than shelling to `git check-ignore`), so the objects are fetched into the trusted checkout through this repository's own `refs/pull/N/head` — never from the contributor's remote — and analysed as data. The working tree stays on the default branch; the workflow contains no `checkout`, `switch` or `worktree` of PR content, asserted by test. The checker is invoked as an argument vector rather than a shell string, and every binding field it receives is re-validated on read.

The uploaded report is still downloaded and identity-bound, and is now compared against the re-derivation. A divergence is reported and **the re-derived result wins** — refusing to comment on a mismatch would let a forged artifact suppress the very finding it was forged to hide.

The adversarial regression test reproduces the attack end to end: a fork PR that introduces `.vscode/tasks.json`, drops `.vscode/` from `.gitignore`, cleans up, and uploads a forged CLEAN report carrying correct identity. It asserts the forged artifact *does* pass identity validation and *would* render CLEAN, then that re-derivation finds 3 findings and renders **BLOCKED** naming the offending commit:

```
::warning::the uploaded artifact reported 0 finding(s) but the trusted checker
re-derived 3; the re-derived result is authoritative
```

Also covered: an honest clean PR still renders CLEAN (re-derivation introduces no false positives), an unusable range errors rather than reading as an empty finding list, and the uploaded artifact is never the file handed to the renderer.

Fork-PR binding, stale-head protection and fail-closed behaviour are unchanged. Suites: 73 / 17 / 32 / 9 new.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: scripts/resolve_repository_integrity_pr.py [hunk 4]
```diff
@@ -0,0 +1,496 @@
...
+SCANNER_WORKFLOW_NAME = "Repository integrity"
+SCANNER_EVENT = "pull_request"
+
+SHA_RE = re.compile(r"^[0-9a-f]{40}$")
+REPO_RE = re.compile(r"^[A-Za-z0-9._-]{1,100}/[A-Za-z0-9._-]{1,100}$")
+# GitHub caps a branch name at 255 bytes. Nothing narrower is imposed: see
+# _branch below for why this value is bounded but not otherwise constrained.
+MAX_BRANCH_LENGTH = 255
+# GitHub logins: alphanumeric and hyphens, 39 characters maximum.
+LOGIN_RE = re.compile(r"^[A-Za-z0-9](?:[A-Za-z0-9-]{0,38})$")
```
[the remark this answers is not in the document]
Comment (TheZupZup) [reply scripts/resolve_repository_integrity_pr.py]: Correct, and a regression from the commit immediately before this one. Fixed in `cb6b1bf`.

Verified before changing anything — the pattern rejected every bot form, and `.github/dependabot.yml` is configured in this repository, so `verify_pull_request` would have raised on every Dependabot PR:

```
dependabot[bot]      -> REJECTED
github-actions[bot]  -> REJECTED
renovate[bot]        -> REJECTED
```

`pr_author` is used for exactly one casefolded comparison, in the checker's `is_external`, and reaches it as an argument-vector element rather than through a shell. It is now held only to what that use needs: a string, a 64-character bound, no ASCII control characters, and no leading hyphen so it cannot be mistaken for an option by the program it is passed to. I confirmed by experiment that a leading hyphen would fail *closed* in argparse rather than inject anything — rejecting it just keeps the argv unambiguous. The same validator now covers `--repo-owner` in the recompute driver.

Tests pin that the three bot logins above bind, and that a leading hyphen, an embedded newline or NUL, an empty string, an over-long value and a non-string are all still refused — on both the binder and the driver.

Worth recording that this and the earlier branch-name finding share a root cause: over-constraining a GitHub-provided value that is only ever used as data, which turns cosmetic hardening into a functional outage for legitimate inputs. Both are now bounded rather than shaped, and I've checked the remaining validators on that basis — the SHA and repository patterns stay strict because both values are placed into API paths, where the shape is load-bearing rather than decorative.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by TheZupZup** [comment]
@codex review

File: .github/workflows/repository-integrity-reporter.yml [hunk 5]
```diff
@@ -22,42 +31,165 @@ jobs:
...
+      # The verdict is therefore re-derived here by the checker on this trusted
+      # default-branch checkout, over the same range, from inputs taken from
+      # the live pull request. Objects are fetched and read; no pull-request
+      # tree is checked out and no pull-request code runs.
+      - name: Re-derive the findings with the trusted checker
+        if: ${{ steps.bind.outputs.post == 'true' }}
+        run: |
+          python3 scripts/recompute_repository_integrity_report.py \
+            --binding "$RUNNER_TEMP/repository-integrity/binding.json" \
+            --checker scripts/check_repository_integrity.py \
```
[the remark this answers is not in the document]
Comment (TheZupZup) [reply .github/workflows/repository-integrity-reporter.yml]: Correct. Fixed in `458fd0a` — the re-derivation now reads the checker from the bound base SHA, as you suggested.

Verified all four premises before changing anything:

1. The scanner does load from the base commit — `git show "$BASE_SHA:$CHECKER_PATH"`.
2. The driver was running `scripts/check_repository_integrity.py` from the reporter's default-branch checkout.
3. The base-commit blob is readable without any checkout (914 lines at current `main`, and it does carry `--json-report`), so binding to it costs nothing structurally.
4. Only `main` exists as a base branch today, so the non-default-branch case is structural rather than currently occurring — but the "checker changes on `main` between scan and report" case applies to `main` PRs too, so the gap is live either way.

Worth being precise about the category: both revisions are maintainer-controlled, so this is a **consistency** requirement rather than a trust one. It matters because a check and the comment that explains it must never disagree — and the CLEAN-while-red direction is the same misleading shape as the forged-CLEAN issue in the neighbouring thread, just reached by version skew instead of forgery.

The decisive new test gives the base commit a *distinguishable* checker revision (a stub emitting a `BASE-REVISION-MARKER` finding) and asserts the re-derived report carries that marker. Had the reporter run its own copy, the marker would be absent — which is exactly how the comment and the check come apart. Also pinned: a base commit carrying no checker refuses rather than reading as clean, and the checker path is constrained against traversal.

The fixtures now commit the checker at their base commit instead of pointing at the reporter's working copy, so every test in that file exercises the real load path rather than a shortcut.

Suites: 73 / 17 / 34 / 14.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: scripts/resolve_repository_integrity_pr.py [hunk 6]
```diff
@@ -0,0 +1,513 @@
...
+    head = _mapping(live.get("head"), "PR head")
+    if binding.head_branch is not None:
+        if _branch(head.get("ref"), "PR head branch") != binding.head_branch:
+            raise BindingError("pull request head branch does not match the scanned head")
+    head_sha = _sha(head.get("sha"), "PR head SHA")
+    # Both are read from the live pull request, so the re-derivation the reporter
+    # performs cannot be steered by anything the scanner run wrote.
+    bound = replace(
+        binding,
+        base_sha=_sha(base.get("sha"), "PR base SHA"),
```
[the remark this answers is not in the document]
Comment (TheZupZup) [reply scripts/resolve_repository_integrity_pr.py:385]: The premise is correct and I confirmed it against this repository, but we're accepting the window rather than closing it. Recorded in `3c9e06e` at the binding site and in the hardening doc.

**`base.sha` is indeed mutable.** PR #513 was created `2026-08-25T06:22:56Z`; commit `9a53243` was committed `2026-08-25T18:59:40Z` — roughly thirteen hours later — and #513 nonetheless reports `9a53243` as its `base.sha`. So it tracks the base branch's current tip, exactly as you describe.

Two corrections to the impact, though, and they're why the trade goes the other way:

**The commit range does not change.** `rev-list head --not base` excludes what is reachable from `base`. Commits pushed to the base *after* the scan are not reachable from a head that predates them, so the same commits are scanned either way. Only the checker *revision* can differ — and only when the push that moved the base also changed `check_repository_integrity.py`, within the seconds between the two workflow runs.

**The proposed remedy has no trusted channel.** "Preserve the scan-time base SHA" means taking it from the artifact, which is contributor-controlled — and a forged `base_sha` pointing at the head would empty the range and produce a CLEAN, which is the forged-CLEAN attack from the neighbouring thread reached through a new field.

It *can* be authenticated: require the claimed base to be an ancestor of the live base tip, which bounds an attacker to real base-branch commits, and picking an older one enlarges the scanned range rather than shrinking it. That is sound. But it would make an artifact field decide something again — the one property this design gives up everywhere else — to close a window that cannot change which commits are examined. Weighed against each other, the narrower risk wins.

So the limitation is documented rather than papered over: what it is, why it is bounded, and what closing it would cost. If the checker's rules ever start changing frequently on `main`, the ancestry-authenticated variant is the fix to reach for.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by TheZupZup** [comment]
@codex address that feedback

Please address the currently unresolved P2 on the latest head: the trusted re-derivation must mirror the scanner's bootstrap checker selection for repository-owner, same-repository PRs whose base either lacks `scripts/check_repository_integrity.py` or predates the `--json-report` interface.

Match the scanner's trust boundary exactly:
- prefer the checker blob from the bound base SHA when present and structured-report capable;
- only allow the bootstrap fallback for a PR authored by the repository owner AND whose head repository is this repository;
- in that bootstrap case, load the checker from the bound PR head SHA as git object data, without checking out or executing any other PR-controlled content;
- external/fork PRs must continue to fail closed when the base checker is missing/too old;
- keep the existing forged-CLEAN protection, fork binding, stale-head protection, base/checker consistency protections, and no `pull_request_target` design.

Add decisive regression tests for both bootstrap cases (missing base checker and pre-`--json-report` checker), plus a negative external/fork case. Run the tooling suites and security checks. Do not merge.

File: scripts/recompute_repository_integrity_report.py [hunk 7]
```diff
@@ -0,0 +1,242 @@
...
+
+    try:
+        if not valid_login(args.repo_owner):
+            raise RecomputeError("invalid repository owner")
+        binding = load_binding(args.binding)
+        args.output.parent.mkdir(parents=True, exist_ok=True)
+        if not args.no_fetch:
+            fetch_objects(binding)
+        checker = load_checker(str(binding["base_sha"]), args.checker_path,
+                               args.output.parent / "trusted-checker.py")
```
[the remark this answers is not in the document]
Comment (TheZupZup) [reply scripts/recompute_repository_integrity_report.py]: Fixed in `c4cb808`: trusted re-derivation now mirrors the scanner's checker selection. It uses the bound base checker normally, and only permits the bound head checker when the PR author is the repository owner and the head repository is the same repository. Fork/external PRs can never use the head checker. Tests cover missing/legacy base checker bootstrap and the negative fork cases.

File: test/tooling/recompute_repository_integrity_report_test.py [hunk 8]
```diff
@@ -0,0 +1,525 @@
...
+            ["BASE-REVISION-MARKER"],
+        )
+
+    def test_external_pr_cannot_bootstrap_when_base_checker_is_missing(self) -> None:
+        """A fork head checker is never executable, even if it is structured."""
+        git(self.repo, "rm", "-q", "scripts/check_repository_integrity.py")
+        git(self.repo, "commit", "-qm", "base without a checker")
+        self.base = git(self.repo, "rev-parse", "HEAD")
+        git(self.repo, "switch", "-qc", "pr")
+        (self.repo / "scripts" / "check_repository_integrity.py").write_bytes(
```
[the remark this answers is not in the document]
Comment (TheZupZup) [reply test/tooling/recompute_repository_integrity_report_test.py:273]: Fixed in `955dfa9`: both missing-checker bootstrap fixtures now recreate `scripts/` after `git rm` removes the last tracked file in that directory. The fresh `Repository integrity` workflow on this head is green, including the guard self-test job.

**Comment by TheZupZup** [comment]
@codex review
